examples/structured_data/deep_neural_decision_forests.py

```diff
@@ -0,0 +1,447 @@
+    "income_bracket",
+]
+
+train_data_url = (
+    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data"
+)
+train_data = pd.read_csv(train_data_url, header=None, names=CSV_HEADER)
+
+test_data_url = (
+    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test"
+)
+test_data = pd.read_csv(test_data_url, header=None, names=CSV_HEADER)
+
+print(f"Train dataset shape: {train_data.shape}")
+print(f"Test dataset shape: {test_data.shape}")
+
+"""
+We remove the first record as it is not a valid data example, and we remove a trailing
+'dot' in the class labels.
+"""
+
+test_data = test_data[1:]
+test_data.income_bracket = test_data.income_bracket.apply(
+    lambda value: value.replace(".", "")
+)
+
+"""
+We store the training and test data splits locally as CSV files.
+"""
+
+train_data_file = "train_data.csv"
+test_data_file = "test_data.csv"
+
+train_data.to_csv(train_data_file, index=False, header=False)
+test_data.to_csv(test_data_file, index=False, header=False)
+
+"""
+## Define dataset metadata
+
+Here, we define the metadata of the dataset that will be useful for reading and parsing
+and encoding input features.
+"""
+
+# A list of the numerical feature names.
+NUMERIC_FEATURE_NAMES = [
+    "age",
+    "education_num",
+    "capital_gain",
+    "capital_loss",
+    "hours_per_week",
+]
+# A dictionary of the categorical features and their vocabulary.
+CATEGORICAL_FEATURES_WITH_VOCABULARY = {
+    "workclass": sorted(list(train_data["workclass"].unique())),
+    "education": sorted(list(train_data["education"].unique())),
+    "marital_status": sorted(list(train_data["marital_status"].unique())),
+    "occupation": sorted(list(train_data["occupation"].unique())),
+    "relationship": sorted(list(train_data["relationship"].unique())),
+    "race": sorted(list(train_data["race"].unique())),
+    "gender": sorted(list(train_data["gender"].unique())),
+    "native_country": sorted(list(train_data["native_country"].unique())),
+}
+# A list of the columns to ignore from the dataset.
+IGNORE_COLUMN_NAMES = ["fnlwgt"]
+# A list of the categorical feature names.
+CATEGORICAL_FEATURE_NAMES = list(CATEGORICAL_FEATURES_WITH_VOCABULARY.keys())
+# A list of all the input features.
+FEATURE_NAMES = NUMERIC_FEATURE_NAMES + CATEGORICAL_FEATURE_NAMES
+# A list of column default values for each feature.
+COLUMN_DEFAULTS = [
+    [0.0] if feature_name in NUMERIC_FEATURE_NAMES + IGNORE_COLUMN_NAMES else ["NA"]
+    for feature_name in CSV_HEADER
+]
+# The name of the target feature.
+TARGET_FEATURE_NAME = "income_bracket"
+# A list of the labels of the target features.
+TARGET_LABELS = [" <=50K", " >50K"]
+
+"""
+## Create `tf.data.Dataset` objects for training and validation
+
+We create an input function to read and parse the file, and convert features and labels
+into a [`tf.data.Dataset`](https://www.tensorflow.org/guide/datasets)
+for training or validation. We also preprocess the input by mapping the target label
```

Comment:
I think this is "training and validation" (separately)

```suggestion
for training and validation. We also preprocess the input by mapping the target label
```